### PR TITLE
Chore: Fix window as any when accessing feature toggles in grafana data

### PR DESCRIPTION
**Pull request by oscarkilhed**
**What is this feature?**

Accessing feature toggles in grafana data is a bit hacky. This tries to at least make it a little bit less hacky after some input @ashharrison90 [here](https://github.com/grafana/grafana/pull/75372#discussion_r1348574882).

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-data/src/utils/featureToggle.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { BootData } from '../types';
+
+declare global {
+  interface Window {
+    grafanaBootData?: BootData;
```
Comment (ashharrison90) [inline packages/grafana-data/src/utils/featureToggle.ts:5]: is this ever not defined? making it required would remove a lot of optional chaining elsewhere 🤔

**Review by ashharrison90 — APPROVED** [review]
looks great! 🙌 
added a minor question

File: packages/grafana-data/src/utils/featureToggle.ts (hunk 1 above)
Comment (oscarkilhed) [reply packages/grafana-data/src/utils/featureToggle.ts:5]: My main concern was that it could be undefined in tests. But maybe I'm being overly cautious.

Comment (ashharrison90) [reply packages/grafana-data/src/utils/featureToggle.ts:5]: @oscarkilhed ah that's a fair concern 👍 

we do already mock out the bootdata in unit tests here: https://github.com/grafana/grafana/blob/oscark/fix-window-as-any-when-accessing-feature-toggles/public/test/jest-setup.ts#L24, so might just be a case of adding a `featureToggles` object here as well

File: packages/grafana-data/src/utils/featureToggle.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  }
+}
+
+/**
+ * Grafana data does not have access to runtime, so we are accessing the window object to get the feature toggles.
```
Comment (academo) [inline packages/grafana-data/src/utils/featureToggle.ts:10]: Could you please rewrite this comment to what this API does and how to use it and not why we have the API?

Consider that other developers (external ones too) will explore the API through this comments and they should be treated as documentation.

File: packages/grafana-data/src/utils/featureToggle.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+ *
+ * @returns Grafana feature toggles object
+ */
+export function getGrafanaFeatureToggles() {
+  return window.grafanaBootData?.settings?.featureToggles;
```
Comment (academo) [inline packages/grafana-data/src/utils/featureToggle.ts:15]: I advice against this. Can you instead create an async api that "waits" for the runtime to be available?

**Review by academo — COMMENTED** [review]
It looks like a good idea to remove the `any` but I rather we don't encourage the usage of global objects. Specially since this global object is supposed to exist only as a mean to start the grafana app and not as an accessible space for other code to read from.

Some alternatives: 

* Is it possible that anything that could be accessing this API will only do it after the grafana app is ready and thus, runtime is populated? if so, can we instead throw an error if something tries to access this API before time? 
* Can we turn this into an async API that reads from runtime.config and waits for it to be populated before returning?


The way it is currently coded, this API is effectively going around the grafana app life cycle

**Review by mckn — COMMENTED** [review]
I do understand why this is done the way we have done it. However I think the long term solution for this particular problem is to move transformations out of `grafana-data` into core. That would remove the need of being able to do this in grafana-data, right?

**Comment by oscarkilhed** [comment]
@mckn Yes, that would definitely solve this issue. Our main issue with moving transformations to core right now is that the code is under Apache 2.0 and most of Grafana is under AGPL. We could perhaps exempt the transformations folder from the AGPL license.

**Comment by mckn** [comment]
> @mckn Yes, that would definitely solve this issue. Our main issue with moving transformations to core right now is that the code is under Apache 2.0 and most of Grafana is under AGPL. We could perhaps exempt the transformations folder from the AGPL license.

hmm, ok ok - some of the transformations already live in core 🙈 So I guess it's a feature with multiple license.

**Pull request closed without merging**